### ISSUE TITLE
Funny JSON handling

### ISSUE THREAD

**Issue by chisholm** [oasis-open/cti-taxii-client#46]
I first started noticing this while working on the user-agent changes (#45), but this isn't directly related to those changes, so I am making a separate issue for this.

One of the things I added (and forgot to mention in the commit message, oops) was to add parameter documentation to _HTTPConnection.post(), which for some reason had been missing.  For the "data" parameter, I just copied text from the [Requests documentation](http://docs.python-requests.org/en/master/api/#requests.Session.post), since it's passed directly into Session.post().  Notice that JSON content is not listed as a legal value.  That was odd, because aren't the TAXII services all JSON-based?  Note also that there is a "json" parameter you can use specifically for JSON.  Wouldn't it have made more sense to use that?  This got me looking at the JSON handling.

The usage of the "data" parameter within taxii2-client is via the Collection.add_objects() method; the actual value is obtained [here](https://github.com/oasis-open/cti-taxii-client/blob/d4fb580379b7ac0ea079cf8becd2348f4e878c9e/taxii2client/__init__.py?#L526-L530):

```python
        if isinstance(bundle, dict):
            if six.PY2:
                bundle = json.dumps(bundle, encoding="utf-8")
            else:
                bundle = json.dumps(bundle)
```

(I just picked master branch head commit as of this writing, to refer to; the aforementioned PR hasn't changed this part.)  This code snip creates JSON which is subsequently passed via "data" parameter, instead of using the "json" parameter and letting Requests take care of the conversion.  There may be some reinvention of the wheel here.  What of that "encoding" parameter though?  Why is it hard-coded, and why make encoding decisions in the Collection endpoint class anyway?

To start with, here's what I think the encoding kwarg actually does: in case there are any binary values in the passed dict, they must be decoded to strings before being encoded as JSON (JSON is a textual format, after all).  The encoding argument describes how to do that.  I.e. the *en*coding argument is actually about *de*coding.

What may have happened is a misunderstanding: that a decision needed to be made here about what to do with high-value codepoints, and the decision was to encode as utf-8.  Lending credence to this idea is that the TAXII 2.0 spec actually defines the string type as `The string data type represents a finite-length string of valid characters from the Unicode coded character set [ISO10646] that are encoded in UTF-8.`  So the spec actually mentions utf-8.  Perhaps the idea was that a json.dumps() call could do double-duty, as a JSON-encode operation followed by a string encode operation.  On the other hand, this code is written to behave differently depending on python version, and there is no usage of encoding on python3.  If encoding needed to happen for python2, why not python3?  On python3 the API changed: there actually is no "encoding" parameter in json.dumps().  Under this misunderstanding, that would mean that the result must remain text, and text is passed to _HTTPConnection.post(), then to Session.post(), which is invalid according to the documentation.  So even if the misunderstanding were true, the code is incorrect.

Also, the "ensure_ascii" param was left at default (True), which means there can be no high-value codepoints in the resulting JSON anyway.  They will all be escaped using the format "\u1234".  JSON specs I looked at do seem to recognize this as an escape syntax, but do we need to do that?  Why not let them pass?

Here's how I think it should work:
- If we actually expect people to pass text-as-binary data in their JSON-serializable objects, we should also allow people to tell us how that text was encoded, not hard-code an assumption.
- The above bullet is probably not worth worrying about; I can't see why anyone would want to do that.  So I don't see any need to specify an encoding there, and have python-version-dependent code.
- Do we really intend for people to be able to pass anything they want to TAXII server endpoints?  If we want to restrict this to JSON, maybe we should be using the "json" kwarg to Session.post(), and mirror that in _HTTPConnection.post().  If want to allow anything, maybe we should mirror Session.post() and allow either "data=" or "json=" as kwargs.  We would only need "json=" in our endpoint classes.  (Btw, when using "json=", Requests automatically encodes the result as utf-8, so it would be compliant.)
- Don't use funny escape syntax unless you really have to.  No point in layering additional escaping which serves no purpose.

**Comment by chisholm** [comment]
I made PR 47 with some cleanup.  I found an even simpler reason the python-version-specific code is incorrect: in python2, the "utf-8" value of the "encoding" kwarg to json.dumps(), is the default!  So really,

```python
if isinstance(bundle, dict):
    if six.PY2:
        bundle = json.dumps(bundle, encoding="utf-8")
    else:
        bundle = json.dumps(bundle)
```

is the same as:

```python
if isinstance(bundle, dict):
    if six.PY2:
        bundle = json.dumps(bundle)
    else:
        bundle = json.dumps(bundle)
```

which is the same as just

```python
if isinstance(bundle, dict):
    bundle = json.dumps(bundle)
```

(Which is in turn incorrect as described above: this still turns `bundle` into a string, which is not a supported type for the `data` kwarg of Session.post().)

Collection.add_objects() now supports dict, strings, and binary.  Some unit tests were added, although unicode support is still poorly tested (e.g. none of the test content includes unicode strings with high-codepoint characters.  I'm not sure how you'd test it anyways, since you can't test how the requests library forms its HTTP requests, to make sure the bundle data you pass in to add_objects() is encoded as expected in the request body... can you?  Or is there something else we should do?).

I changed _HTTPConnection.post() to support both "json=" and "data=" kwargs, so its interface is similar to Session.post().  I opted not to use the "json=" kwarg though in Collection.add_objects(), for dict bundles, because although it works, you have no control over how the JSON is formed.  The requests library just does a maximally-simple 'json.dumps()', which does the "\uXXXX" escaping of high-codepoint characters, and this behavior would be inconsistent with what happens when you pass strings, which would produce a plain utf-8 encoding with no escaping.  So "json=" is currently unused, but might be handy for end-users or other purposes...?

This issue and PR is really just some ideas and illustration of the problem and my solution.  Feedback welcome :)

**Comment by chisholm** [comment]
An addendum to this: there was apparently already some discussion and fixes related to this, in issue #22, which I hadn't known about.  But the fix implemented there didn't work.  Apparently _HTTPConnection.post() originally had the "json=" kwarg, and it was changed.  There wasn't a clear rationale given in that issue; perhaps it was thought that people would try to pass python-stix2 objects and expect that they would be automatically serialized to JSON.  As mentioned above, I wound up not using "json=" either, so it would be easy to drop support for it.  I just thought maybe someone might find it handy someday.
